parity/run.rs

```diff
@@ -1002,3 +1024,32 @@ fn wait_for_exit(
 	let _ = exit.1.wait(&mut l);
 	l.clone()
 }
+
+fn wait_for_drop<T>(w: Weak<T>) {
+	let sleep_duration = Duration::from_secs(1);
+	let warn_timeout = Duration::from_secs(30);
+	let max_timeout = Duration::from_secs(120);
+
+	let instant = Instant::now();
+	let mut warned = false;
+
+	loop {
+		{
+			if let None = w.upgrade() {
+				break;
+			}
+		}
+
+		if instant.elapsed() > max_timeout {
```

Comment:
Nitpick. Imho this loop is more concise and cleaner for me (but don't have strong opinion):
```rust
while instance.elapsed() < max_timeout {
   if w.upgrade().is_none() {
     return;
   }
   if !warned && instant.elapsed() > warn_timeout {
     warned = true;
     warn!(...);
   }
}
warn('unclean');
```